### PR TITLE
MINOR: Add Processing Guarantees in Streams docs

### PR DESCRIPTION
**Pull request by guozhangwang**

### CODE REVIEW

**Comment by guozhangwang** [comment]
@mjsax @apurvam @hachikuji @sriramsub for reviews.

File: docs/streams.html [hunk 1]
```diff
@@ -158,6 +161,27 @@
         </p>
         <br>
 
+        <h2><a id="streams_processing_guarantee" href="#streams_processing_guarantee">Processing Guarantees</a></h2>
+
+        <p>
+            In stream processing, one of the most frequently asked question is "does my stream processing system guarantee that each record is processed once and only once, even if some failures are encountered in the middle of processing?"
+            Failing to guarantee exactly-once stream processing is a deal-breaker for many applications that cannot tolerate any data-loss or date duplicates, and in that case a batch-oriented framework is usually used in addition
+            to the stream processing pipeline, known as the <a href="http://lambda-architecture.net/">Lambda Architecture</a>.
+            Prior to 0.11.0.0, Kafka only provides at-least-once delivery guarantees and hence any stream processing systems that leverages it as the backend storage cannot guarantee end-to-end exactly-once semantics.
```
Comment (mjsax) [inline docs/streams.html]: `cannot` -> `could not`

File: docs/streams.html [hunk 2]
```diff
@@ -158,6 +161,27 @@
...
+            and Kafka Streams has hence added the end-to-end exactly-once processing semantics by leveraging these features.
+            More specifically, it guarantees that for any record read from the source Kafka topics, its processing results will be reflected exactly once in the output Kafka topic as well as in the updated processing state.
+            Note the key difference between Kafka Streams end-to-end exactly-once guarantee with other stream processing frameworks' claimed guarantees is that Kafka Streams tightly integrates with the underlying Kafka storage system and ensure that
+            committing on the input topic offsets, updating on the processing state, and producing to the output topics will be completed atomically instead of treating Kafka as an external system that may have side-effects.
+            To read more details on how this is done inside Kafka Streams, readers are recommended to read <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-129%3A+Streams+Exactly-Once+Semantics">this KIP</a>.
```
Comment (mjsax) [inline docs/streams.html]: `this KIP` -> `KIP-129`

File: docs/streams.html [hunk 3]
```diff
@@ -158,6 +161,27 @@
...
+
+            In order to achieve exactly-once semantics when running Kafka Streams applications, users can simply set the <code>processing.guarantee</code> config value to <b>exactly_once</b> (default value is <b>at_least_once</b>).
+            More details can be found <a href="#streamsconfigs">here</a>
```
Comment (mjsax) [inline docs/streams.html]: `here` -> `in section "Kafka Streams Configs"`

**Review by mjsax — CHANGES_REQUESTED** [review]

File: docs/streams.html [hunk 4]
```diff
@@ -158,6 +161,27 @@
         </p>
         <br>
 
+        <h2><a id="streams_processing_guarantee" href="#streams_processing_guarantee">Processing Guarantees</a></h2>
+
+        <p>
+            In stream processing, one of the most frequently asked question is "does my stream processing system guarantee that each record is processed once and only once, even if some failures are encountered in the middle of processing?"
+            Failing to guarantee exactly-once stream processing is a deal-breaker for many applications that cannot tolerate any data-loss or date duplicates, and in that case a batch-oriented framework is usually used in addition
```
Comment (apurvam) [inline docs/streams.html]: "date duplicates" -> "data duplicates"

File: docs/streams.html [hunk 5]
```diff
@@ -46,19 +46,22 @@
         <h2><a id="streams_overview" href="#streams_overview">Overview</a></h2>
 
         <p>
-        Kafka Streams is a client library for processing and analyzing data stored in Kafka and either write the resulting data back to Kafka or send the final output to an external system. It builds upon important stream processing concepts such as properly distinguishing between event time and processing time, windowing support, and simple yet efficient management of application state.
+            Kafka Streams is a client library for processing and analyzing data stored in Kafka and either write the resulting data back to Kafka or send the final output to an external system.
```
Comment (apurvam) [inline docs/streams.html]: "either write ... or send " -> "either writes .. or sends" (note the plural).

File: docs/streams.html [hunk 6]
```diff
@@ -46,19 +46,22 @@
...
         <p>
-        Some highlights of Kafka Streams:
+            Some highlights of Kafka Streams:
         </p>
 
         <ul>
             <li>Designed as a <b>simple and lightweight client library</b>, which can be easily embedded in any Java application and integrated with any existing packaging, deployment and operational tools that users have for their streaming applications.</li>
             <li>Has <b>no external dependencies on systems other than Apache Kafka itself</b> as the internal messaging layer; notably, it uses Kafka's partitioning model to horizontally scale processing while maintaining strong ordering guarantees.</li>
             <li>Supports <b>fault-tolerant local state</b>, which enables very fast and efficient stateful operations like windowed joins and aggregations.</li>
+            <li>Enables <b>exactly-once</b> processing semantics to guarantee that each record will be processed once and only once even when there is a failure in the middle of processing.</li>
```
Comment (apurvam) [inline docs/streams.html]: Enables -> Supports ?

File: docs/documentation.html [hunk 7]
```diff
@@ -71,7 +71,7 @@
 
     <h2><a id="streams" href="/0102/documentation/streams">9. Kafka Streams</a></h2>
     <p>
-        Kafka Streams is a client library for processing and analyzing data stored in Kafka and either write the resulting data back to Kafka or send the final output to an external system. It builds upon important stream processing concepts such as properly distinguishing between event time and processing time, windowing support, and simple yet efficient management of application state.
+        Kafka Streams is a client library for processing and analyzing data stored in Kafka and either write the resulting data back to Kafka or send the final output to an external system. It builds upon important stream processing concepts such as properly distinguishing between event time and processing time, windowing support, exactly-once processing semantics and simple yet efficient management of application state.
```
Comment (apurvam) [inline docs/documentation.html]: "either write ... or send " -> "either writes .. or sends" (note the plural).

File: docs/streams.html [hunk 8]
```diff
@@ -158,6 +161,27 @@
...
+            Prior to 0.11.0.0, Kafka only provides at-least-once delivery guarantees and hence any stream processing systems that leverages it as the backend storage could not guarantee end-to-end exactly-once semantics.
+            In fact, even for those stream processing systems that claim to support exactly-once processing, as long as they are reading from / writing to Kafka as the source / sink, their applications cannot actually guarantee that
+            no duplicates will be generated throughout the pipeline.
+
+            Since the 0.11.0.0 release, Kafka has added support to allow its producers to send messages to different topic partitions in a <a href="https://kafka.apache.org/documentation/#semantics">transactional and idempotent manner</a>,
+            and Kafka Streams has hence added the end-to-end exactly-once processing semantics by leveraging these features.
+            More specifically, it guarantees that for any record read from the source Kafka topics, its processing results will be reflected exactly once in the output Kafka topic as well as in the updated processing state.
```
Comment (apurvam) [inline docs/streams.html]: Should 'updated processing state' simply be 'state store' ? Or do you mean something else by this?

File: docs/streams.html [hunk 9]
```diff
@@ -158,6 +161,27 @@
...
+            Note the key difference between Kafka Streams end-to-end exactly-once guarantee with other stream processing frameworks' claimed guarantees is that Kafka Streams tightly integrates with the underlying Kafka storage system and ensure that
+            committing on the input topic offsets, updating on the processing state, and producing to the output topics will be completed atomically instead of treating Kafka as an external system that may have side-effects.
```
Comment (apurvam) [inline docs/streams.html]: I would reword this passage :  

"and ensures that commits on the input topic offsets, updates on the state store, and writes to the output topics will be completed atomically instead of treating Kafka as an external system that may have side-effects."

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5360/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5375/
Test PASSed (JDK 7 and Scala 2.11).

File: docs/streams.html (hunk 5 above)
Comment (guozhangwang) [reply docs/streams.html]: ack.

File: docs/streams.html (hunk 6 above)
Comment (guozhangwang) [reply docs/streams.html]: Was trying to use different verbs for each line :P but ack.

File: docs/streams.html (hunk 4 above)
Comment (guozhangwang) [reply docs/streams.html]: ack.

File: docs/streams.html (hunk 8 above)
Comment (guozhangwang) [reply docs/streams.html]: ack.

File: docs/streams.html (hunk 9 above)
Comment (guozhangwang) [reply docs/streams.html]: Makes sense.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5425/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5410/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by ijuma** [comment]
cc @hachikuji

File: docs/documentation.html [hunk 10]
```diff
@@ -71,7 +71,7 @@
 
     <h2><a id="streams" href="/0102/documentation/streams">9. Kafka Streams</a></h2>
     <p>
-        Kafka Streams is a client library for processing and analyzing data stored in Kafka and either write the resulting data back to Kafka or send the final output to an external system. It builds upon important stream processing concepts such as properly distinguishing between event time and processing time, windowing support, and simple yet efficient management of application state.
+        Kafka Streams is a client library for processing and analyzing data stored in Kafka and either writes the resulting data back to Kafka or sends the final output to an external system. It builds upon important stream processing concepts such as properly distinguishing between event time and processing time, windowing support, exactly-once processing semantics and simple yet efficient management of application state.
```
Comment (hachikuji) [inline docs/documentation.html]: Can we take out "and either writes the resulting data back to Kafka or sends the final output to an external system"? Seems unnecessary for the high-level summary.

File: docs/streams.html [hunk 11]
```diff
@@ -46,19 +46,22 @@
         <h2><a id="streams_overview" href="#streams_overview">Overview</a></h2>
 
         <p>
-        Kafka Streams is a client library for processing and analyzing data stored in Kafka and either write the resulting data back to Kafka or send the final output to an external system. It builds upon important stream processing concepts such as properly distinguishing between event time and processing time, windowing support, and simple yet efficient management of application state.
+            Kafka Streams is a client library for processing and analyzing data stored in Kafka and either writes the resulting data back to Kafka or sends the final output to an external system.
```
Comment (hachikuji) [inline docs/streams.html]: Maybe split this into two sentences:
> Kafka Streams is a client library for processing and analyzing data stored in Kafka. Results can either be written back into Kafka or to an external system.

File: docs/streams.html [hunk 12]
```diff
@@ -46,19 +46,22 @@
...
         <p>
-        Some highlights of Kafka Streams:
+            Some highlights of Kafka Streams:
         </p>
 
         <ul>
             <li>Designed as a <b>simple and lightweight client library</b>, which can be easily embedded in any Java application and integrated with any existing packaging, deployment and operational tools that users have for their streaming applications.</li>
             <li>Has <b>no external dependencies on systems other than Apache Kafka itself</b> as the internal messaging layer; notably, it uses Kafka's partitioning model to horizontally scale processing while maintaining strong ordering guarantees.</li>
             <li>Supports <b>fault-tolerant local state</b>, which enables very fast and efficient stateful operations like windowed joins and aggregations.</li>
+            <li>Supports <b>exactly-once</b> processing semantics to guarantee that each record will be processed once and only once even when there is a failure in the middle of processing.</li>
```
Comment (hachikuji) [inline docs/streams.html]: Maybe mention both client and broker failures are included?

File: docs/streams.html [hunk 13]
```diff
@@ -158,6 +161,27 @@
         </p>
         <br>
 
+        <h2><a id="streams_processing_guarantee" href="#streams_processing_guarantee">Processing Guarantees</a></h2>
+
+        <p>
+            In stream processing, one of the most frequently asked question is "does my stream processing system guarantee that each record is processed once and only once, even if some failures are encountered in the middle of processing?"
+            Failing to guarantee exactly-once stream processing is a deal-breaker for many applications that cannot tolerate any data-loss or data duplicates, and in that case a batch-oriented framework is usually used in addition
+            to the stream processing pipeline, known as the <a href="http://lambda-architecture.net/">Lambda Architecture</a>.
+            Prior to 0.11.0.0, Kafka only provides at-least-once delivery guarantees and hence any stream processing systems that leverages it as the backend storage could not guarantee end-to-end exactly-once semantics.
```
Comment (hachikuji) [inline docs/streams.html]: "leverages" -> "leverage" (since "stream processing systems" is plural)

**Review by hachikuji — COMMENTED** [review]
Thanks for the patch. Left a few minor comments.

File: docs/streams.html (hunk 11 above)
Comment (guozhangwang) [reply docs/streams.html]: Thinking about this, I think it is Okay to just remove the line as your first suggestion since it is not necessary as part of a high-level description.

**Comment by guozhangwang** [comment]
@hachikuji addressed your comments. Also bumped up version variables from 0102 to 0110.

File: docs/documentation.html [hunk 14]
```diff
@@ -69,15 +69,15 @@
...
+        Kafka Streams is a client library for processing and analyzing data stored in Kafka. It builds upon important stream processing concepts such as properly distinguishing between event time and processing time, windowing support, exactly-once processing semantics and simple yet efficient management of application state.
     </p>
     <p>
         Kafka Streams has a <b>low barrier to entry</b>: You can quickly write and run a small-scale proof-of-concept on a single machine; and you only need to run additional instances of your application on multiple machines to scale up to high-volume production workloads. Kafka Streams transparently handles the load balancing of multiple instances of the same application by leveraging Kafka's parallelism model.
     </p>
 
-    <p>Learn More about Kafka Streams read <a href="/0102/documentation/streams">this</a> Section.</p>
+    <p>Learn More about Kafka Streams read <a href="/0110/documentation/streams">this</a> Section.</p>
```
Comment (hachikuji) [inline docs/documentation.html:80]: Hmm... Should we be doing the same trick that we do for the other includes? The nice thing is that we no longer have to update the versions.

Comment (guozhangwang) [reply docs/documentation.html:80]: @derrickdoo and I have been discussing about this, and the conclusion is that in the entry page `documentation.html` it is much trickier to do and hence we did not do that.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5514/
Test PASSed (JDK 8 and Scala 2.12).

File: docs/documentation.html (hunk 14 above)
Comment (hachikuji) [reply docs/documentation.html:80]: Haha, ok. I'll take you and Derrick's word for it.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5528/
Test PASSed (JDK 7 and Scala 2.11).

**Review by hachikuji — APPROVED** [review]
LGTM

**Comment by hachikuji** [comment]
Merging to trunk and 0.11.0.
